src/main/scala/fr/acinq/bitcoin/Psbt.scala

```diff
@@ -25,22 +25,70 @@ case class Psbt(global: Psbt.Global, inputs: Seq[Psbt.Input], outputs: Seq[Psbt.
-                         outputIndex: Int,
+  def updateWitnessInput(outPoint: OutPoint,
+                         txOut: TxOut,
                          redeemScript: Option[Seq[ScriptElt]] = None,
                          witnessScript: Option[Seq[ScriptElt]] = None,
                          sighashType: Option[Int] = None,
                          derivationPaths: Map[PublicKey, KeyPathWithMaster] = Map.empty): Try[Psbt] = Try {
+    val inputIndex = global.tx.txIn.indexWhere(_.outPoint == outPoint)
+    require(inputIndex >= 0, "psbt transaction does not spend the provided outpoint")
+    val updatedInput = inputs(inputIndex) match {
+      case input: PartiallySignedInputWithoutUtxo => PartiallySignedWitnessInput(
+        txOut,
+        None,
+        sighashType.orElse(input.sighashType),
+        Map.empty,
+        derivationPaths ++ input.derivationPaths,
+        redeemScript,
+        witnessScript,
+        input.ripemd160,
+        input.sha256,
+        input.hash160,
+        input.hash256,
+        input.unknown
+      )
+      case input: PartiallySignedWitnessInput => input.copy(
+        txOut = txOut,
+        redeemScript = redeemScript.orElse(input.redeemScript),
+        witnessScript = witnessScript.orElse(input.witnessScript),
+        sighashType = sighashType.orElse(input.sighashType),
+        derivationPaths = input.derivationPaths ++ derivationPaths
+      )
+      case _: PartiallySignedNonWitnessInput => return Failure(new IllegalArgumentException("cannot update segwit input: it has already been updated with non-segwit data"))
+      case _: FinalizedInput => return Failure(new IllegalArgumentException("cannot update segwit input: it has already been finalized"))
+    }
+    this.copy(inputs = inputs.updated(inputIndex, updatedInput))
+  }
+
+  /**
+   * Implements the PSBT updater role; adds information about a given segwit utxo.
+   * Note that we always fill the nonWitnessUtxo (see https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#cite_note-7).
+   *
+   * @param inputTx         transaction containing the utxo.
+   * @param outputIndex     index of the utxo in the inputTx.
+   * @param redeemScript    redeem script if known and applicable (when using p2sh-embedded segwit).
+   * @param witnessScript   witness script if known and applicable (when using p2wsh).
+   * @param sighashType     sighash type if one should be specified.
+   * @param derivationPaths derivation paths for keys used by this utxo.
+   * @return psbt with the matching input updated.
+   */
+  def updateWitnessInputTx(inputTx: Transaction,
+                           outputIndex: Int,
+                           redeemScript: Option[Seq[ScriptElt]] = None,
+                           witnessScript: Option[Seq[ScriptElt]] = None,
+                           sighashType: Option[Int] = None,
+                           derivationPaths: Map[PublicKey, KeyPathWithMaster] = Map.empty): Try[Psbt] = Try {
```

Comment:
why not have `updateWitnessInputTx()` call `updateWitnessInput()` ? 
and why not overload ``updateWitnessInput()` to take either a tx or an outpoint + txout ?

Reply:
Right! I didn't overload because we can't overload when we provide default arguments (annoying scala limitation).
But I could change it to take an `Either` instead of using two separate functions, WDYT?

Reply:
Ah I had forgotten about that limitation, 2 different methods is fine then